File: components/unit-security/src/main/java/org/eclipse/dirigible/components/security/domain/Constraints.java
/*
 * Copyright (c) 2022 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-FileCopyrightText: 2022 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
 * SPDX-License-Identifier: EPL-2.0
 */
package org.eclipse.dirigible.components.security.domain;

import java.util.ArrayList;
import java.util.List;

import com.google.gson.annotations.Expose;

/**
 * The Class Constraints.
 */
public class Constraints {

    /** The constraints. */
	@Expose
    private List<Constraint> constraints = new ArrayList<>();

    /**
     * Builds the security accesses.
     *
     * @param location the location
     * @return the list
     */
    public List<Access> buildSecurityAccesses(String location) {
        List<Access> securityAccesses = new ArrayList<>();

        Integer keyIndex = 1;

        for (Constraint securityAccessConstraint : constraints) {
            for (String role : securityAccessConstraint.getRoles()) {
                Access securityAccess = new Access();
                securityAccess.setLocation(location);
                securityAccess.setType(Access.ARTEFACT_TYPE);
                securityAccess.setName(keyIndex.toString());
                securityAccess.updateKey();
                securityAccess.setMethod(securityAccessConstraint.getMethod());
                securityAccess.setRole(role);
                securityAccess.setPath(securityAccessConstraint.getPath());
                securityAccess.setScope(securityAccessConstraint.getScope());

                securityAccesses.add(securityAccess);
                keyIndex++;
            }
        }

        return securityAccesses;
    }

	@Override
	public String toString() {
		return "Constraints [constraints=" + constraints + "]";
	}
    
    
}

File: components/ide/ide-ui-documents/src/main/resources/META-INF/dirigible/ide-documents/api/constraints.js
/*
 * Copyright (c) 2023 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
 * SPDX-License-Identifier: EPL-2.0
 */
import { rs } from "sdk/http";
import * as constraintsProcessor from "./processors/constraintsProcessor";

rs.service()
    .resource("")
    .get(function (ctx, request, response) {
        if (request.isUserInRole("Operator")) {
            let accessDefinitions = constraintsProcessor.getAccessDefinitions();
            response.println(JSON.stringify(accessDefinitions, null, 2));
        } else {
            response.setStatus(response.FORBIDDEN);
            response.println("Access forbidden");
        }
    })
    .put(function (ctx, request, response) {
        if (request.isUserInRole("Operator")) {
            let accessDefinitions = request.getJSON();
            constraintsProcessor.updateAccessDefinitions(accessDefinitions);
            response.println(JSON.stringify(accessDefinitions, null, 2));
        } else {
            response.setStatus(response.FORBIDDEN);
            response.println("Access forbidden");
        }
    })
    .execute();